content/blog/2024/2024-05-15-java-functional-interfaces.md

```diff
@@ -0,0 +1,2028 @@
+| Reference to a Constructor                              | Refers to a class constructor, allowing us to create new instances. This type is useful when we need to create objects without explicitly calling a constructor. | <code>Supplier<Car> carSupplier = Car::new;<code>|
+
+## Predicates
+
+**Predicates are functional interfaces in Java that represent boolean-valued functions of a single argument.** They are commonly used for filtering, testing, and conditional operations.
+
+The `Predicate` functional interface is part of the `java.util.function` package and defines a functional method `test(T t)` that returns a `boolean`.  It also provides default methods that allow combine two predicates.
+
+```java
+@FunctionalInterface
+public interface Predicate<T> {
+    boolean test(T t);
+    // default methods
+}
+```
+
+The `test()` method evaluates the predicate on the input argument and determines whether it satisfies the condition defined by the predicate.
+
+We often use predicates with the `stream()` API for filtering elements based on certain conditions. Pass them as arguments to methods like `filter()` to specify the criteria for selecting elements from a collection.
+
+Let's see filtering in action:
+
+```java
+public class PredicateTest {
+  @Test
+  void testFiltering() {
+    List<Integer> numbers = List.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+    Predicate<Integer> isEven = num -> num % 2 == 0;
+    
+    List<Integer> actual = numbers.stream().filter(isEven).toList();
+    List<Integer> expected = List.of(2, 4, 6, 8, 10);
+    
+    Assertions.assertEquals(expected, actual);
+  }
+}
+```
+
+In the test `testFiltering()` method, first we populate a list of integers. Then we define a predicate `isEven` to check if a number is even. Using `stream()` and `filter()` methods, we filter the list to contain only even numbers. Finally, we compare the filtered list to expected list.
+
+### Combining Predicates
+
+**We can combine predicates using logical operators such as `and()`, `or()`, `negate()` and `not()` to create complex conditions.**
+
+Let's see how to combine the practices:
+
+```java
+@Test
+void testPredicate() {
+  List<Integer> numbers = List.of(-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5);
+  Predicate<Integer> isZero = num -> num == 0;
+  Predicate<Integer> isPositive = num -> num > 0;
+  Predicate<Integer> isOdd = num -> num % 2 == 1;
+
+  Predicate<Integer> isPositiveOrZero = isPositive.or(isZero);
+  Predicate<Integer> isPositiveAndOdd = isPositive.and(isOdd);
+  Predicate<Integer> isNotPositive = Predicate.not(isPositive);
+  Predicate<Integer> isNotZero = isZero.negate();
+  Predicate<Integer> isAlsoZero = isPositive.negate().and(isNegative.negate());
```

Comment:
The predicate `isNegative` isn't defined.

Reply:
Fixed